engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java

```diff
@@ -0,0 +1,574 @@
+      "otherInterop.isString(otherString)"
+  }, limit = "3")
+  boolean equalsStrings(Object selfString, Object otherString,
+      @CachedLibrary("selfString") InteropLibrary selfInterop,
+      @CachedLibrary("otherString") InteropLibrary otherInterop) {
+    String selfJavaString;
+    String otherJavaString;
+    try {
+      selfJavaString = selfInterop.asString(selfString);
+      otherJavaString = otherInterop.asString(otherString);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+    return Normalizer.compare(
+        selfJavaString,
+        otherJavaString,
+        Normalizer.FOLD_CASE_DEFAULT
+    ) == 0;
+  }
+
+  @Specialization(guards = {
+      "selfInterop.hasArrayElements(selfArray)",
+      "otherInterop.hasArrayElements(otherArray)"
+  }, limit = "3")
+  boolean equalsArrays(Object selfArray, Object otherArray,
+      @CachedLibrary("selfArray") InteropLibrary selfInterop,
+      @CachedLibrary("otherArray") InteropLibrary otherInterop,
+      @Cached EqualsAnyNode equalsNode
+      ) {
+    try {
+      long selfSize = selfInterop.getArraySize(selfArray);
+      if (selfSize != otherInterop.getArraySize(otherArray)) {
+        return false;
+      }
+      for (long i = 0; i < selfSize; i++) {
+        Object selfElem = selfInterop.readArrayElement(selfArray, i);
+        Object otherElem = otherInterop.readArrayElement(otherArray, i);
+        if (!equalsNode.execute(selfElem, otherElem)) {
+          return false;
+        }
+      }
+      return true;
+    } catch (UnsupportedMessageException | InvalidArrayIndexException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /** Equals for Atoms and AtomConstructors */
+
+  @Specialization
+  boolean equalsAtomConstructors(AtomConstructor selfConstructor, AtomConstructor otherConstructor) {
+    return selfConstructor == otherConstructor;
+  }
+
+  /**
+   * How many {@link EqualsAnyNode} should be created for fields in specialization for atoms.
+   */
+  static final int equalsNodeCountForFields = 10;
+
+  static EqualsAnyNode[] createEqualsNodes(int size) {
+    EqualsAnyNode[] nodes = new EqualsAnyNode[size];
+    Arrays.fill(nodes, EqualsAnyNode.build());
+    return nodes;
+  }
+
+  @Specialization
+  boolean equalsAtoms(Atom self, Atom other,
+      @Cached LoopConditionProfile loopProfile,
+      @Cached(value = "createEqualsNodes(equalsNodeCountForFields)", allowUncached = true) EqualsAnyNode[] fieldEqualsNodes,
+      @Cached InvokeEqualsNode atomInvokeEqualsNode,
+      @Cached ConditionProfile enoughEqualNodesForFieldsProfile,
+      @Cached ConditionProfile constructorsNotEqualProfile) {
+    if (constructorsNotEqualProfile.profile(
+        self.getConstructor() != other.getConstructor()
+    )) {
+      return false;
+    }
+    assert self.getFields().length == other.getFields().length;
+
+    if (atomOverridesEquals(self)) {
+      return atomInvokeEqualsNode.execute(self, other);
+    }
```

Comment:
As we were discussing, I think the plan was to move the override to the top, right?

My intuition tells me that this should probably be a separate specialization guarded by `atomOverridesEquals(self)`. But regardless, I think it has to be separate as the idea was that for these cases, `other` may not even be an atom, right?

Reply:
> As we were discussing, I think the plan was to move the override to the top, right?

No. The plan is to provide custom equality and hash_code via comparator - see https://www.pivotaltracker.com/story/show/183945328.

> other may not even be an atom, right?

No, `other` must be atom. That will be handled by the `comparator` static method - since it is *static*, we can always be sure that we will compare values of the same type.

Reply:
Okay, still the test I proposed in a comment below will not work right now. For now we are still using `==` overriding, so why not change the order to make it correct?

Anyway, I'd add the test there so that we mark that we want to allow this kind of behaviour, even if it has to be marked as pending until the new approach to equality is implemented.

Reply:
Ohh, right. I am sorry - yes, it is perfectly OK to have different constructors. I will move this `atomOverridesEquals(self)` check above same constructors check.

Reply:
Thanks!